### ISSUE TITLE
Better speed with VPN connection

### ISSUE THREAD

**Issue by Daisenki** [streamlink/streamlink-twitch-gui#526]
Hi,

I'm using Twitch with a VPN and try to find a good setting with the program + MPV. I already know it's running much better than the normal Twitch with the browser but I have a few questions and can't find all answers at the wikis / google...

Right now I have a MPV cache with 1sec and 250-500KiB as example at standard input / HLS live 3 and threads 2... is it good to increase the HLS threads for a stable stream? Should I switch to HLS pass through?

Has it any effect to open any ports for the traffic? I can see at the log files ports are used from streamlink-twitch-gui TCP:5915; Python TCP:5917-5921 and TCP:6162

**Comment by bastimeyer** [comment]
You should ask those questions on the Streamlink repo. Streamlink Twitch GUI is just a Twitch browser and a **launcher and configurator for Streamlink**, which means that it has nothing to do with downloading, buffering or caching of the streams.

If you're having questions regarding Streamlink, you should consult its documentation:
https://streamlink.github.io/cli.html#cmdoption-hls-live-edge
https://streamlink.github.io/cli.html#cmdoption-hls-segment-threads

Let me try to help you anyway...

> is it good to increase the HLS threads for a stable stream?

HLS streams are a sequence of stream segments/chunks split by the streaming server. Twitch is splitting their streams in segments of two seconds, which results in chunk sizes of up to ~1.5MiB when the stream is using a bitrate of 6Mbit/s. Watching the re-encoded lower quality streams will ofc reduce these sizes.

Increasing the value of the `--hls-segment-threads` parameter depends on your connection speed and on the value of `--hls-live-edge`, which you've set to `3`. This `--hls-live-edge` parameter makes Streamlink download and buffer the most X recent chunks from the current *live* position of the stream until they are - if completed - written to the player's stdin pipe (passed to the player process). Increasing this value will move you further back from the live position, but will give your connection more time to finish downloading the chunks. If you're increasing the download threads, more HLS chunks are being downloaded in parallel, but this comes at the cost of your shared connection speed between all downloads and some chunks, especially the one that's required next, might not get downloaded quickly enough because of that. So increasing this value might result in a worse buffering experience.

> Should I switch to HLS pass through?

https://github.com/streamlink/streamlink-twitch-gui/wiki/Streamlink-configuration#hls-passthrough
If you're using this method, then you'll need to configure your player and its own HLS implementation instead, because it'll completely bypass Streamlink.

> Has it any effect to open any ports for the traffic?

I don't know why this is a concern here. Each HTTP client, namely Streamlink Twitch GUI and Streamlink (Python), need to communicate with the Twitch servers via simple HTTP(s) requests. Streamlink Twitch GUI needs to query the Twitch API and download preview images from the CDN, Streamlink needs to query the Twitch API and needs to download the HLS playlists and chunks. Depending on the `--hls-segment-threads` configuration, it will use multiple HTTP connections. Streamlink Twitch GUI doesn't have any control over these HTTP requests because they are made by Chromium.

**Comment by Daisenki** [comment]
Thanks for your information. I need to test more with my connection. With more threads there is a little bit more speed, sometimes it drops to zero speed and the stream stops, its better with increasing the edge live from default 3 to 5, so I guess speed isn't the problem. 

I have seen with "Standard input" there is the slowest speed... but I don't know why? I can get with HTTP and HLS till 2-3mbit download speed BUT the stream is not stable. With "Standard input" there is only 400-600kib with the same settings.

Open ports was only a idea, sometimes it give more traffic speed for programs but unfortunately the streaming port is always random so it would a lot of work to open it.

**Comment by bastimeyer** [comment]
If you're not using "Standard input" as player input method and "HTTP" instead, you're telling Streamlink to launch an internal HTTP server from which the player then reads. This internal HTTP server is serving a *progressive stream* of *glued together* HLS chunks. This method is only required for players which don't support reading from their `stdin` pipe (some players on Windows), but makes no difference otherwise. However, by choosing HTTP, the software you're using for analyzing your network traffic is probably being confused about this, because it's seeing this local traffic as well.
https://github.com/streamlink/streamlink-twitch-gui/wiki/Streamlink-configuration#player-input

There is no reason for Streamlink to behave differently in terms of stream download speeds when choosing different player input methods. Well, except HLS, but I've already told you that Streamlink won't do anything when you're using this method.

**Comment by Daisenki** [comment]
The page https://streamlink.github.io/issues.html was helpful too because with the setting cache for MPV I have now a good speed and no buffer anymore. Thanks for your help bastimeyer!
